File: riveredge-frontend/src/apps/kuaicaiwu/services/finance/partnerStatement.ts
import { apiRequest } from '../../../../services/api';

export interface PartnerStatementLine {
  date: string;
  doc_type: string;
  doc_code: string;
  summary?: string;
  debit: number;
  credit: number;
  balance: number;
}

export interface PartnerStatementSummary {
  opening_balance: number;
  debit_total: number;
  credit_total: number;
  closing_balance: number;
}

export interface PartnerStatementPreview {
  partner_id: number;
  partner_name: string;
  partner_type: string;
  start_date: string;
  end_date: string;
  company_name: string;
  balance_label: string;
  summary: PartnerStatementSummary;
  lines: PartnerStatementLine[];
  partner_snapshot?: Record<string, unknown>;
}

export interface PartnerStatement {
  id: number;
  statement_code: string;
  partner_id: number;
  partner_name: string;
  partner_type: string;
  statement_period: string;
  start_date: string;
  end_date: string;
  opening_balance: number;
  debit_total: number;
  credit_total: number;
  closing_balance: number;
  status: string;
  company_name?: string;
  transaction_details?: {
    summary?: PartnerStatementSummary;
    lines?: PartnerStatementLine[];
    balance_label?: string;
    partner_snapshot?: Record<string, unknown>;
  };
  confirmed_at?: string;
  sent_at?: string;
  sent_channel?: string;
  dispute_reason?: string;
  disputed_at?: string;
  notes?: string;
  created_at: string;
}

const API = '/apps/kuaicaiwu/partner-statements';

export const partnerStatementService = {
  preview: (params: {
    partner_id: number;
    partner_type: string;
    start_date: string;
    end_date: string;
  }) =>
    apiRequest<PartnerStatementPreview>(`${API}/preview`, { method: 'GET', params }),

  create: (data: {
    partner_id: number;
    partner_type: string;
    statement_period: string;
    start_date?: string;
    end_date?: string;
    notes?: string;
    attachments?: Array<{ uid?: string; name?: string; status?: string; url?: string }>;
  }) => apiRequest<PartnerStatement>(API, { method: 'POST', data }),

  list: (params?: Record<string, unknown>) =>
    apiRequest<{ items: PartnerStatement[]; total: number }>(API, { method: 'GET', params }),

  get: (id: number) => apiRequest<PartnerStatement>(`${API}/${id}`, { method: 'GET' }),

  confirm: (id: number) => apiRequest<PartnerStatement>(`${API}/${id}/confirm`, { method: 'POST' }),

  markSent: (id: number, data: { channel: string; notes?: string }) =>
    apiRequest<PartnerStatement>(`${API}/${id}/mark-sent`, { method: 'POST', data }),

  dispute: (id: number, reason: string) =>
    apiRequest<PartnerStatement>(`${API}/${id}/dispute`, { method: 'POST', data: { reason } }),

  delete: (id: number) => apiRequest<void>(`${API}/${id}`, { method: 'DELETE' }),

  exportFile: (id: number, format: 'xlsx' | 'pdf') =>
    apiRequest<Blob>(`${API}/${id}/export`, {
      method: 'GET',
      params: { format },
      responseType: 'blob',
    }),
};

export const PARTNER_STATEMENT_STATUS_MAP: Record<string, { text: string; color: string }> = {
  Draft: { text: '草稿', color: 'default' },
  Confirmed: { text: '已确认', color: 'processing' },
  Sent: { text: '已发送', color: 'success' },
  Disputed: { text: '有异议', color: 'warning' },
};

export const SENT_CHANNEL_OPTIONS = [
  { label: '导出后微信/邮件发送', value: 'wechat_manual' },
  { label: '打印邮寄', value: 'print' },
  { label: '导出文件发送', value: 'export' },
  { label: '邮件手动发送', value: 'email_manual' },
  { label: '其他', value: 'other' },
];

export function downloadBlob(blob: Blob, filename: string) {
  const url = URL.createObjectURL(blob);
  const a = document.createElement('a');
  a.href = url;
  a.download = filename;
  a.click();
  URL.revokeObjectURL(url);
}

File: riveredge-backend/src/apps/kuaicaiwu/models/partner_statement.py
"""
往来对账单模型
"""

from tortoise import fields
from core.models.base import BaseModel


class PartnerStatement(BaseModel):
    """
    往来对账单 (Partner Statement)

    用于记录与客户或供应商在特定时间段内的所有往来交易摘要，作为双方对账的正式凭据。
    包含：期初余额、本期发生额、期末余额。
    """

    class Meta:
        table = "apps_kuaicaiwu_partner_statements"
        table_description = "管理会计 - 往来对账单"
        indexes = [
            ("tenant_id", "partner_id"),
            ("statement_period",),
            ("status",),
        ]

    id = fields.IntField(pk=True, description="主键ID")
    tenant_id = fields.IntField(description="租户ID")
    statement_code = fields.CharField(max_length=50, unique=True, description="对账单号")

    partner_id = fields.IntField(description="往来单位ID")
    partner_name = fields.CharField(max_length=200, description="往来单位名称")
    partner_type = fields.CharField(max_length=20, description="单位类型 (Customer/Supplier)")

    statement_period = fields.CharField(max_length=20, description="对账周期 (YYYY-MM)")
    start_date = fields.DateField(description="开始日期")
    end_date = fields.DateField(description="结束日期")

    opening_balance = fields.DecimalField(max_digits=14, decimal_places=2, description="期初余额")
    debit_total = fields.DecimalField(max_digits=14, decimal_places=2, description="本期借方总额")
    credit_total = fields.DecimalField(max_digits=14, decimal_places=2, description="本期贷方总额")
    closing_balance = fields.DecimalField(max_digits=14, decimal_places=2, description="期末余额")

    status = fields.CharField(max_length=20, default="Draft", description="状态 (Draft/Confirmed/Sent/Disputed)")

    transaction_details = fields.JSONField(null=True, description="往来明细快照")
    company_name = fields.CharField(max_length=200, null=True, description="我司名称快照")

    confirmed_at = fields.DatetimeField(null=True, description="内部确认时间")
    confirmed_by = fields.IntField(null=True, description="确认人ID")

    sent_at = fields.DatetimeField(null=True, description="发送时间")
    sent_by = fields.IntField(null=True, description="发送操作人ID")
    sent_channel = fields.CharField(max_length=30, null=True, description="发送渠道")

    dispute_reason = fields.TextField(null=True, description="异议说明")
    disputed_at = fields.DatetimeField(null=True, description="异议记录时间")

    notes = fields.TextField(null=True, description="备注")
    attachments = fields.JSONField(null=True, description="附件列表")
    created_by = fields.IntField(null=True, description="创建人ID")
    created_at = fields.DatetimeField(auto_now_add=True, description="创建时间")
    deleted_at = fields.DatetimeField(null=True, description="删除时间")

    def __str__(self):
        return f"Statement: {self.partner_name} - {self.statement_period}"
